### PR TITLE
Feature/38723 notification settings frontend

### PR DESCRIPTION
**Pull request by b12f**
https://community.openproject.org/projects/openproject/work_packages/38723/activity

### CODE REVIEW

File: config/locales/js-en.yml [hunk 1]
```diff
@@ -580,20 +576,32 @@ en:
         text_update_date: "%{date} by"
         total_count_warning: "Showing the %{newest_count} most recent notifications. %{more_count} more are not displayed."
       settings:
-        default_all_projects: 'Default for all projects'
```
Comment (HDinger) [inline config/locales/js-en.yml:586]: This string is still used [here](https://github.com/opf/openproject/blob/feature/38723-notification-settings-frontend/frontend/src/app/features/user-preferences/notifications-settings/row/notification-setting-row.component.ts#L40). Same applies for some of the other strings that were removed here.
I am aware that this component is not used any longer, so I'd argue to either remove the component as well or re-add the strings again.

File: spec/features/notifications/digest_mail_spec.rb [hunk 2]
```diff
@@ -56,49 +56,6 @@
   end
 
   it 'sends a digest mail based on the configuration', with_settings: { journal_aggregation_time_minutes: 0 } do
-    # Configure the digest
-    notification_settings_page.visit!
-
-    notification_settings_page.expect_setting channel: :mail_digest,
```
Comment (HDinger) [inline spec/features/notifications/digest_mail_spec.rb:66]: I think it would be better not to remove this here, and really test if changes to the SettingsPage also lead to changes in the digest. 
Otherwise you test here whether the mail is correct with predefined settings, and below whether changes on the page update the settings. But what is missing is the part that changes made on the settings page then directly lead to the mail being sent with the changed settings (and not still with the old ones).

File: frontend/src/app/shared/components/table/table.sass [hunk 3]
```diff
@@ -0,0 +1,19 @@
+.op-table
```
Comment (HDinger) [inline frontend/src/app/shared/components/table/table.sass:1]: I am not really happy with that styles being being part of a globally named "table" block. From a BEM block named `op-table` I would expect to reflect how (at least) most of the tables in the system look like. And I think it is very unlikely that in the future all of our tables should look like this. 
However I currently also lack a good alternative, maybe we should shortly brainstorm about it 🤷‍♀️

**Review by HDinger — CHANGES_REQUESTED** [review]
When looking at the page I noticed several differences the to specified design in Figma:

### General
* The sub-headlines are way to large. They are even large then the global toolbar
Bildschirmfoto 2021-09-08 um 07 57 29

* In Chrome, it looks like the labels and the checkboxes are not correctly aligned (probably because there is so much space above). In Firefox it looks better. 
Bildschirmfoto 2021-09-08 um 07 59 18

* In Safari mobile, the checkbox labels are not wrapped or truncated. They are simply cut off. I'd prefer adding an overflow rule anyway in case there is a really long word in any translated language.

### The table
* The text "Notify me" in the first cell missing
* The projects should be linked
Bildschirmfoto 2021-09-08 um 08 02 05

* The link to remove a project setting should look different (and have a different wording)

* Same applies for the button to add a project setting

* When clicking the button to add a project setting, there is no space between the table and the select box. Further the select box is way to large.
Bildschirmfoto 2021-09-08 um 08 05 40

* When having many projects I cannot scroll the table and thus don't see all of the content any more.

File: config/locales/js-en.yml (hunk 1 above)
Comment (b12f) [reply config/locales/js-en.yml:586]: Ah yeah I forgot to remove those files

File: spec/features/notifications/digest_mail_spec.rb (hunk 2 above)
Comment (b12f) [reply spec/features/notifications/digest_mail_spec.rb:66]: That is basically tested in the other spec however, and email digests will soon be removed which is why I talked to @ulferts and we came up with this solution. I can put those tests back in however.

File: frontend/src/app/shared/components/table/table.sass (hunk 3 above)
Comment (b12f) [reply frontend/src/app/shared/components/table/table.sass:1]: I understand. I think this is mostly something that design should decide. Are these styles going to be default from now on or not

File: frontend/src/app/shared/components/forms/checkbox-field/checkbox-field.component.html [hunk 4]
```diff
@@ -0,0 +1,45 @@
+<ng-container *ngIf="!hidden">
+  <label class="op-form-field--label-wrap op-checkbox-field--label-wrap">
```
Comment (HDinger) [inline frontend/src/app/shared/components/forms/checkbox-field/checkbox-field.component.html:2]: Out of curiosity: I totally see the benefit of having an angular component, but why did you create your own HTML-tree (and classes) instead of using at least some of the class we already have? None of the the required designs is really new. We already have checkboxes with labels on the right side and an explanatory text below. You can have a look here: http://localhost:3000/assets/styleguide.html#forms

**Review by HDinger — CHANGES_REQUESTED** [review]
* There are still some differences between the text that was specified and the text that is now displayed:
  * The text below "Assigned to me or accountable" is wrong
  * In the table, the explanatory text for "Assigned to me or accountable" and "I am @mentioned" is missing
  * In the table, the text to restore the project settings is wrong. Further the icon is missing.

* When I add a new project to the list, it is added at the very end. However, when I save, the order is changed (I think to an alphabetical sorting). As user, I expect the order to remain unchanged after save. At least I thought at first, that the setting was not saved at all.

* There are still some alignment issues with the checkboxes. In Chrome, the normal check boxes look good, but in Firefox they are a bit off
**Chrome**
Bildschirmfoto 2021-09-14 um 13 02 04

**Firefox**
Bildschirmfoto 2021-09-14 um 13 01 42

Further those with additional text below are not correctly aligned in neither of the browsers.

* Please add a meaningful text for the browser tab as it is one of the more easily achievable accessibility fixes we can provide. Currently it is only written "OpenProject".

* Finally, the spacings don't match the specified ones. For example the space between the checkboxes and the horizontal line as well as the space below the line or the space to the table. This is an issue for nearly every spacing on the page.
Bildschirmfoto 2021-09-14 um 13 06 57

File: frontend/src/app/shared/components/forms/checkbox-field/checkbox-field.component.html (hunk 4 above)
Comment (b12f) [reply frontend/src/app/shared/components/forms/checkbox-field/checkbox-field.component.html:2]: `op-form-field`, `op-form`, and `op-option-list` were created for the add user modal, where for the first time we had a set of related style changes:

* Label above the field instead of next to it
* Validation errors are shown inline

I didn't (and don't) think it would've been a great idea to e.g. have a modifier on the old `form--field` styles, because the existing styles are not in standard BEM style and really hard to refactor. (`frontend/src/global_styles/content/_forms.sass` is 900 lines of sass) They work well in the current usecases, but I figured instead of trying to put changes in there potentially breaking some far away forms, I'd create components specific to the new styles.

These were then adapted to work with the formly dynamic forms that were used on the project settings. Here `op-form` was also finalized a tad more.

Here I created `op-checkbox-field` because I didn't want to blow up `op-form-field` with more checkbox-specific changes. `op-form-field` expects the input to be below the label and some other specialties, however in this design the checkboxes came before the text. As you can see, `op-checkbox-field` still expects to be based on an `op-form-field`, but overwrites and rearranges the children https://github.com/opf/openproject/pull/9636/files/f3aa8efa4fa7dc138fad2c7fb7f4e86362ea7b8f#diff-9a31b6533922f486f0af10580880dd636098f43d91c4620047a12cec442d0eb3R3

This is a lot more than what used to be there for checkboxes; https://github.com/opf/openproject/pull/9636/files/f3aa8efa4fa7dc138fad2c7fb7f4e86362ea7b8f#diff-d79513a013f1fa2f0bcc4d8f12b3faff7180f047665fdd4d8d15527a2842c521L54 which was code that expected the same HTML structure as for `op-form-field`. However ,the two HTML structures now slightly differ.

tl;dr Instead of bloating`form--field` with `op-form-field` and  `op-form-field` with more versions I opted for this. I'd be okay with saying, `op-checkbox-field` (the BEM block) should be a modifier of `op-form-field`, but I wouldn't try to adapt the old styles to the new layouts or add more styles modifiers there.

**Comment by b12f** [comment]
Re. the text at the bottom of the tables; this was changed: https://community.openproject.org/projects/openproject/work_packages/38671/activity#activity-13?#activity-13

Re. the spacing between checkboxes and their labels; I've now centered the checkboxes vertically next to the text. This is also not to spec however. I'm not sure how much pixel pushing we want to do to accommodate different browsers and how they handle text renders, especially without an official spec w/r/t spacings and typography.

Re. Spacings around the rest of the form I've now applied the `op-form` block, which I think handles the spacings pretty nicely. Without a reusable spec I also wouldn't put too much effort into making these spacings exactly like the figma file.

File: frontend/src/app/shared/components/forms/checkbox-field/checkbox-field.component.html (hunk 4 above)
Comment (HDinger) [reply frontend/src/app/shared/components/forms/checkbox-field/checkbox-field.component.html:2]: That makes totally sense for me. Thanks for the detailed explanation 👍
